relay-server/src/actors/envelopes.rs

```diff
@@ -458,176 +467,132 @@ impl Handler<HandleEnvelope> for EnvelopeManager {
+                    }))
                 }
             })
-            .and_then(clone!(envelope_context, |(
-                envelope,
-                project_state,
-                sampling_project_state,
-            )| {
-                let message = ProcessEnvelope {
-                    envelope,
-                    project_state,
-                    sampling_project_state,
-                    start_time,
-                    scoping: envelope_context.borrow().scoping(),
-                };
-
+            .and_then(move |process_message| {
                 processor
-                    .send_tracked(message, *envelope_context.borrow())
-                    .map_err(|_err| ProcessingError::ScheduleFailed)
+                    .send(process_message)
+                    .map_err(|_| ProcessingError::ScheduleFailed)
                     .flatten()
-            }))
-            .and_then(clone!(envelope_context, |processed| {
-                let project_cache = ProjectCache::from_registry();
-                let rate_limits = processed.rate_limits;
-
+            })
+            .and_then(move |processed| {
                 // Processing returned new rate limits. Cache them on the project to avoid expensive
                 // processing while the limit is active.
+                let rate_limits = processed.rate_limits;
                 if rate_limits.is_limited() {
+                    let project_cache = ProjectCache::from_registry();
                     project_cache.do_send(UpdateRateLimits::new(project_key, rate_limits));
                 }
 
-                match processed.envelope {
-                    Some(envelope) => {
-                        envelope_context.borrow_mut().update(&envelope);
-                        Ok(envelope)
-                    }
-                    None => Err(ProcessingError::RateLimited),
-                }
-            }))
+                processed.envelope.ok_or(ProcessingError::RateLimited)
+            })
             .into_actor(self)
-            .and_then(move |envelope, slf, _| {
-                let scoping = envelope_context.borrow().scoping();
+            .and_then(move |(envelope, mut envelope_context), slf, _| {
+                let scoping = envelope_context.scoping();
                 slf.send_envelope(project_key, envelope, scoping, start_time)
-                    .then(move |result| {
-                        result.map_err(|error| {
-                            let envelope_context = envelope_context.borrow();
+                    .then(move |result| match result {
+                        Ok(_) => {
+                            envelope_context.accept();
+                            Ok(())
+                        }
+                        Err(error) => {
                             let outcome = Outcome::Invalid(DiscardReason::Internal);
 
-                            match error {
+                            Err(match error {
```

Comment:
totally a nitpicking: but I dislike round brackets spanning this many lines, I'd always assign this and then put `Err()` around the variable on a next line.

Reply:
Makes sense. This entire match statement should probably move to a function, or could be simplified by changing the error type to contain the SendError. 

Since this is going to move soon anyway, I'll play with the best version of this in the follow-up.